src/workerd/api/eventsource.c++

```diff
@@ -0,0 +1,520 @@
+               "This ReadableStream is locked");
+  JSG_REQUIRE(!readable->isDisturbed(), TypeError,
+              "This ReadableStream has already been read from");
+  auto eventsource = jsg::alloc<EventSource>(js);
+  eventsource->run(js, kj::mv(readable), false /* No reconnection attempts */);
+  return kj::mv(eventsource);
+}
+
+EventSource::EventSource(jsg::Lock& js, jsg::Url url, kj::Maybe<EventSourceInit> init)
+    : context(IoContext::current()),
+      impl({
+        .url = kj::mv(url),
+        .options = kj::mv(init).orDefault({}),
+      }),
+      abortController(jsg::alloc<AbortController>()),
+      readyState(State::CONNECTING) {}
+
+EventSource::EventSource(jsg::Lock& js)
+    : context(IoContext::current()),
+      abortController(jsg::alloc<AbortController>()),
+      readyState(State::CONNECTING) {}
+
+void EventSource::notifyError(jsg::Lock& js, const jsg::JsValue& error, bool reconnecting) {
+  if (readyState == State::CLOSED) return;
+
+  // Abort the connection if it hasn't already been. This will be a non-op if the
+  // controller has already been aborted.
+  abortController->abort(js, error);
+
+  if (!reconnecting) readyState = State::CLOSED;
+  else readyState = State::CONNECTING;
+
+  // Dispatch the error event.
+  dispatchEventImpl(js, jsg::alloc<ErrorEvent>(js, error));
+
+  // Log the error as an uncaught exception for debugging purposes.
+  IoContext::current().logUncaughtException(UncaughtExceptionSource::ASYNC_TASK, error);
+}
+
+void EventSource::notifyOpen(jsg::Lock& js) {
+  if (readyState == State::CLOSED) return;
+  readyState = State::OPEN;
+  dispatchEventImpl(js, jsg::alloc<OpenEvent>());
+}
+
+void EventSource::notifyMessages(jsg::Lock& js, kj::Array<PendingMessage> messages) {
+  if (readyState == State::CLOSED) return;
+  js.tryCatch([&] {
+    for (auto& message : messages) {
+        auto data = kj::str(kj::delimited(kj::mv(message.data), "\n"_kjc));
+        if (data.size() == 0) continue;
+        dispatchEventImpl(js, jsg::alloc<MessageEvent>(
+            kj::mv(message.event),
+            kj::mv(data),
+            kj::mv(message.id),
+            impl.map([](FetchImpl& i) -> jsg::Url& { return i.url; })));
+    }
+  }, [&](jsg::Value exception) {
+    // If we end up with an exception being thrown in one of the event handlers, we will
+    // stop trying to process the messages and instead just error the EventSource.
+    notifyError(js, jsg::JsValue(exception.getHandle(js)));
+  });
+}
+
+void EventSource::reconnect(jsg::Lock& js) {
+  KJ_ASSERT(impl != kj::none);
+  readyState = State::CONNECTING;
+  abortController = jsg::alloc<AbortController>();
+  auto signal = abortController->getSignal();
+  context.awaitIo(js, signal->wrap(context.afterLimitTimeout(reconnectionTime)))
+      .then(js, JSG_VISITABLE_LAMBDA((self=JSG_THIS), (self), (jsg::Lock& js) mutable {
+    self->start(js);
+  }), JSG_VISITABLE_LAMBDA((self=JSG_THIS),(self),(jsg::Lock& js, jsg::Value exception) {
```

Comment:
Syntax seems a little off here.

Reply:
Not too worried about the formatting here.